planner/core/common_plans.go

```diff
@@ -591,6 +600,115 @@ REBUILD:
 	return err
 }
 
+var (
+	heapDumpInterval = time.Minute * 3
+	heapDumpTime     = time.Now()
+	debugThreshold   = 0.6 // 60%
+)
+
+func debugPlanCacheInfo(sctx sessionctx.Context) {
+	// step 1: dump plan cache info: tot-keys, tot-vals, top10-vals
+	logPlanCacheInfo(sctx)
+
+	// step 2: dump heap if the current memory usage is larger than 60%.
+	memTot, err := memory.MemTotal()
+	if err != nil {
+		logutil.BgLogger().Warn("[Plan-Cache-Patch] get total memory usage error", zap.Error(err))
+		return
+	}
+	memUse, err := memory.MemUsed()
+	if err != nil {
+		logutil.BgLogger().Warn("[Plan-Cache-Patch] get used memory usage error", zap.Error(err))
+		return
+	}
+	if memTot == 0 || memUse == 0 {
+		logutil.BgLogger().Warn("[Plan-Cache-Patch] invalid memory usage values",
+			zap.Uint64("mem-tot", memTot), zap.Uint64("mem-use", memUse))
+		return
+	}
+	if float64(memUse)/float64(memTot) < debugThreshold {
+		return
+	}
+	dumpHeap(sctx)
+}
+
+type intHeap []int
+
+func (h intHeap) Len() int            { return len(h) }
+func (h intHeap) Less(i, j int) bool  { return h[i] > h[j] }
+func (h intHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
+func (h *intHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
+func (h *intHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+func logPlanCacheInfo(sctx sessionctx.Context) {
+	c := sctx.PreparedPlanCache()
+	var totVals, totKeys int
+	top10 := new(intHeap)
+	heap.Init(top10)
+
+	c.ForEach(func(k kvcache.Key, v kvcache.Value) {
+		numVals := len(v.([]*PlanCacheValue))
+		totVals += numVals
+		totKeys += 1
+		heap.Push(top10, numVals)
+		if top10.Len() > 10 {
+			top10.Pop()
+		}
```

Comment:
How this part can expression the meaning for top10? It's only add the numVals to the slice and if the length of slice is more than 10, it pop the end of the element in the slice.

Reply:
Yes, but it's a max heap, the minimum value is popped each time, and finally, the maximum 10 values will be left in the end.